### PR TITLE
fix(pgwire): allow configuring tcp keepalive idle time

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

It's common for an SQL statement not returning any data from the pgwire server for a long time (typically when executing DDL). If the connection is proxied through a load balancer, the LB may consider the connection idle and drop it unexpectedly.

To resolve this issue, we may set the tcp keepalive idle time to a lower value. Here we take a reference to the value in AWS ELB, which is 350 seconds. Note that this can be done in any place dealing with the connection, but doing it on the pgwire server eliminates the need for any additional configuration.

The value is configurable through CLI args and defaults to 300 seconds.

----

FYI, the default keepalive idle time is decided by the operating system, typically 2 hours.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by chenzl25 — APPROVED** [review]
LGTM. It seems to be related to the DBT EOF error.

File: src/utils/pgwire/src/net.rs [hunk 1]
```diff
@@ -79,6 +80,13 @@ impl Listener {
             Self::Tcp(listener) => {
                 let (stream, addr) = listener.accept().await?;
                 stream.set_nodelay(true)?;
+                // Set TCP keepalive to 5 minutes, which is less than the connection idle timeout of 350 seconds in AWS ELB.
+                // https://docs.aws.amazon.com/elasticloadbalancing/latest/network/network-load-balancers.html#connection-idle-timeout
+                {
+                    let r = socket2::SockRef::from(&stream);
+                    let ka = socket2::TcpKeepalive::new().with_time(Duration::from_secs(300));
+                    r.set_tcp_keepalive(&ka)?;
+                }
```
Comment (stdrc) [inline src/utils/pgwire/src/net.rs:89]: Interesting. Is this a hack or a common usage of `socket2`?

Comment (BugenZhao) [reply src/utils/pgwire/src/net.rs:90]: I think it's the common usage and the only way for setting these extra parameters until https://github.com/tokio-rs/tokio/issues/3082 is addressed.

**Comment by stdrc** [comment]
What about cp to 2.0?

**Review by stdrc — APPROVED** [review]

File: src/utils/pgwire/src/net.rs [hunk 2]
```diff
@@ -79,6 +80,14 @@ impl Listener {
             Self::Tcp(listener) => {
                 let (stream, addr) = listener.accept().await?;
                 stream.set_nodelay(true)?;
+                // Set TCP keepalive to 5 minutes, which is less than the connection idle timeout of 350 seconds in AWS ELB.
+                // https://docs.aws.amazon.com/elasticloadbalancing/latest/network/network-load-balancers.html#connection-idle-timeout
+                #[cfg(not(madsim))]
+                {
+                    let r = socket2::SockRef::from(&stream);
+                    let ka = socket2::TcpKeepalive::new().with_time(Duration::from_secs(300));
```
Comment (arkbriar) [inline src/utils/pgwire/src/net.rs:88]: Shall we make it a configurable value in case there's another default timeout in other environments?

Comment (BugenZhao) [reply src/utils/pgwire/src/net.rs]: Done

**Review by arkbriar — APPROVED** [review]
LGTM!

**Pull request merged**